diff --git a/renderdoc/driver/d3d12/d3d12_device.cpp b/renderdoc/driver/d3d12/d3d12_device.cpp
index ded4cf66e..e5cec2d09 100644
--- a/renderdoc/driver/d3d12/d3d12_device.cpp
+++ b/renderdoc/driver/d3d12/d3d12_device.cpp
@@ -566,6 +566,7 @@ WrappedID3D12Device::WrappedID3D12Device(ID3D12Device *realDevice, D3D12InitPara
   m_pDevice13 = NULL;
   m_pDevice14 = NULL;
   m_pDownlevel = NULL;
+  m_pSwapChain = NULL;
   if(m_pDevice)
   {
     m_pDevice->QueryInterface(__uuidof(ID3D12Device1), (void **)&m_pDevice1);
@@ -932,6 +933,8 @@ WrappedID3D12Device::~WrappedID3D12Device()
 
   SAFE_DELETE(m_ResourceManager);
 
+  SAFE_RELEASE(m_pSwapChain);
+
   SAFE_RELEASE(m_DRED.m_pReal);
   SAFE_RELEASE(m_DRED.m_pReal1);
   SAFE_RELEASE(m_DREDSettings.m_pReal);
@@ -1755,6 +1758,8 @@ bool WrappedID3D12Device::Serialise_WrapSwapchainBuffer(SerialiserType &ser, IDX
 
   SERIALISE_CHECK_READ_ERRORS();
 
+  // TODO, add real swapChain?
+
   if(IsReplayingAndReading())
   {
     ID3D12Resource *fakeBB = NULL;
@@ -1768,6 +1773,52 @@ bool WrappedID3D12Device::Serialise_WrapSwapchainBuffer(SerialiserType &ser, IDX
 
     HRESULT hr = S_OK;
 
+    if(m_UseRealSwapchain)
+    {
+      if(NULL == m_pSwapChain)
+      {
+        HWND hwnd = (HWND)Window::Create(0, 0, (uint32_t)BackbufferDescriptor.Width,
+                                         BackbufferDescriptor.Height, L"replayWindow", true);
+        if(hwnd)
+        {
+          IDXGIFactory7 *dxgiFactory = NULL;
+
+          typedef HRESULT(WINAPI * PFN_CREATE_DXGI_FACTORY)(UINT, REFIID, void **);
+
+          PFN_CREATE_DXGI_FACTORY CreateDXGIFactory2Func = (PFN_CREATE_DXGI_FACTORY)GetProcAddress(
+              GetModuleHandleA("dxgi.dll"), "CreateDXGIFactory2");
+
+          hr = CreateDXGIFactory2Func(DXGI_CREATE_FACTORY_DEBUG, IID_PPV_ARGS(&dxgiFactory));
+          if(FAILED(hr))
+          {
+            RDCERR("Create DxgiFactory Fail");
+            return false;
+          }
+
+          DXGI_SWAP_CHAIN_DESC1 swapchainDesc1 = {};
+          swapchainDesc1.Width = (uint32_t)BackbufferDescriptor.Width;
+          swapchainDesc1.Height = BackbufferDescriptor.Height;
+          swapchainDesc1.Format = SwapbufferFormat;
+          swapchainDesc1.SampleDesc.Count = 1;
+          swapchainDesc1.SampleDesc.Quality = 0;
+          swapchainDesc1.BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+          swapchainDesc1.BufferCount = 3;
+          swapchainDesc1.SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD;
+
+          hr = dxgiFactory->CreateSwapChainForHwnd(m_Queue->GetReal(), hwnd, &swapchainDesc1,
+                                                   nullptr, nullptr, &m_pSwapChain);
+          if(FAILED(hr))
+          {
+            RDCERR("Create Swapchain Fail");
+            dxgiFactory->Release();
+            return false;
+          }
+
+          dxgiFactory->Release();
+        }
+      }
+    }
+
     D3D12_HEAP_PROPERTIES heapProps;
     heapProps.Type = D3D12_HEAP_TYPE_DEFAULT;
     heapProps.CPUPageProperty = D3D12_CPU_PAGE_PROPERTY_UNKNOWN;
@@ -2485,6 +2536,7 @@ HRESULT WrappedID3D12Device::Present(ID3D12GraphicsCommandList *pOverlayCommandL
 
   RenderDoc::Inst().AddActiveDriver(RDCDriver::D3D12, true);
 
+  static uint32_t s_endFrameNumber = 0;
   // serialise the present call, even for inactive windows
   if(IsActiveCapturing(m_State))
   {
@@ -2501,6 +2553,10 @@ HRESULT WrappedID3D12Device::Present(ID3D12GraphicsCommandList *pOverlayCommandL
     Serialise_Present(ser, backbuffer, SyncInterval, Flags);
 
     m_FrameCaptureRecord->AddChunk(scope.Get());
+    if(m_FrameCounter < s_endFrameNumber)
+    {
+      return S_OK;
+    }
   }
 
   if(!activeWindow)
@@ -2527,6 +2583,8 @@ HRESULT WrappedID3D12Device::Present(ID3D12GraphicsCommandList *pOverlayCommandL
 
     m_AppControlledCapture = false;
     m_CapturedFrames.back().frameNumber = m_FrameCounter;
+
+    s_endFrameNumber = m_FrameCounter + 3;
   }
 
   return S_OK;
@@ -2561,6 +2619,11 @@ bool WrappedID3D12Device::Serialise_Present(SerialiserType &ser, ID3D12Resource
     action.copyDestination = PresentedBackbuffer;
 
     cmd.AddAction(action);
+
+    if(m_UseRealSwapchain && (m_pSwapChain != NULL))
+    {
+      m_pSwapChain->Present(0, 0);
+    }
   }
 
   return true;
